src/main/java/io/aiven/kafka/connect/s3/S3SinkTask.java

```diff
@@ -46,40 +74,150 @@ public S3SinkTask() {
     @Override
     public void start(final Map<String, String> props) {
         Objects.requireNonNull(props, "props hasn't been set");
-        this.config = new S3SinkConfig(props);
-        this.streamMap = new S3StreamWriter(config);
+        config = new S3SinkConfig(props);
+
+        final var awsEndpointConfig = newEndpointConfiguration(this.config);
+        final var s3ClientBuilder =
+            AmazonS3ClientBuilder
+                .standard()
+                .withCredentials(
+                    new AWSStaticCredentialsProvider(
+                        new BasicAWSCredentials(
+                            config.getAwsAccessKeyId().value(),
+                            config.getAwsSecretKey().value()
+                        )
+                    )
+                );
+        if (Objects.isNull(awsEndpointConfig)) {
+            s3ClientBuilder.withRegion(config.getAwsS3Region());
+        } else {
+            s3ClientBuilder.withEndpointConfiguration(awsEndpointConfig).withPathStyleAccessEnabled(true);
+        }
+        s3Client = s3ClientBuilder.build();
+        outputWriter = getOutputWriter();
+
+        try {
+            recordGrouper = RecordGrouperFactory.newRecordGrouper(config);
+        } catch (final Exception e) {
+            throw new ConnectException("Unsupported file name template " + config.getFilename(), e);
+        }
+    }
+
+    private OutputWriter getOutputWriter() {
+        switch (this.config.getFormatType()) {
+            case CSV:
+                return OutputPlainWriter.builder().addFields(config.getOutputFields()).build();
+            case JSONL:
+                return JsonLinesOutputWriter.builder().addFields(config.getOutputFields()).build();
+            default:
+                throw new ConnectException("Unsupported format type " + config.getFormatType());
+        }
     }
 
     @Override
     public void put(final Collection<SinkRecord> records) throws ConnectException {
+        Objects.requireNonNull(records, "records cannot be null");
         LOGGER.info("Processing {} records", records.size());
-        records.stream()
-            .map(r -> Pair.of(new TopicPartition(r.topic(), r.kafkaPartition()), r))
-            .forEach(p -> streamMap.write(p.getLeft(), p.getRight()));
+        records.forEach(record -> recordGrouper.put(record));
     }
 
     @Override
     public void flush(final Map<TopicPartition, OffsetAndMetadata> offsets) {
-        offsets.keySet().forEach(streamMap::flush);
+        recordGrouper.records().forEach(this::flushFile);
+        recordGrouper.clear();
     }
 
-    @Override
-    public void close(final Collection<TopicPartition> partitions) throws ConnectException {
-        partitions.forEach(tp -> {
-            LOGGER.info("Close {}", tp);
-            streamMap.close(tp);
-        });
+    private void flushFile(final String filename, final List<SinkRecord> records) {
```

Comment:
`filename` (that comes from the record grouper) is not used, but should be. `newStreamFor` creates the filename anew without taking templating into account.

Reply:
Good point. If this is true, there should be a tests to cover that